File: framework/core/compiler/lexer.js
import { CompileError, ErrorCode, throwError } from './errors.js'
import { createPos, createLoc } from './ast.js'

export const TokenType = {
    TAG_OPEN: 'TAG_OPEN',
    TAG_OPEN_CLOSE: 'TAG_OPEN_CLOSE',
    TAG_CLOSE: 'TAG_CLOSE',
    TAG_SELF_CLOSE: 'TAG_SELF_CLOSE',
    COMMENT: 'COMMENT',
    ATTR_NAME: 'ATTR_NAME',
    ATTR_EQUALS: 'ATTR_EQUALS',
    ATTR_VALUE: 'ATTR_VALUE',
    TEXT: 'TEXT',
    MUSTACHE_OPEN: 'MUSTACHE_OPEN',
    MUSTACHE_EXPR: 'MUSTACHE_EXPR',
    MUSTACHE_CLOSE: 'MUSTACHE_CLOSE',
    EOF: 'EOF'
}

export function tokenize(source) {
    const lexer = new Lexer(source)
    return lexer.tokenize()
}

class Lexer {
    constructor(source) {
        this.source = source
        this.tokens = []
        this.pos = 0
        this.line = 1
        this.col = 1
        this._mode = 'text'
    }

    tokenize() {
        while (!this._eof()) {
            if (this._mode === 'text') {
                this._readText()
            } else if (this._mode === 'tag') {
                this._readInTag()
            }
        }
        this._push(TokenType.EOF, '', this._startPos())
        return this.tokens
    }

    _readText() {
        if (this._match('<!--')) {
            this._readComment()
            return
        }

        if (this._match('</')) {
            this._readClosingTag()
            return
        }

        if (this._peek() === '<' && this._isTagNameStart(this._peekAt(1))) {
            this._readOpenTag()
            return
        }

        if (this._match('{{')) {
            this._readMustache()
            return
        }

        const start = this._startPos()
        let text = ''
        while (!this._eof()) {
            if ((this._peek() === '<') || this._lookingAt('{{')) break
            text += this._advance()
        }

        if (text) this._push(TokenType.TEXT, text, start)
    }

    _readOpenTag() {
        const start = this._startPos()
        this._consume('<')
        const name = this._readTagName()
        this._pushAt(TokenType.TAG_OPEN, name, start)
        this._mode = 'tag'
    }

    _readClosingTag() {
        const start = this._startPos()
        this._consume('<')
        this._consume('/')
        const name = this._readTagName()
        this._pushAt(TokenType.TAG_OPEN_CLOSE, name, start)
        this._skipWhitespace()
        if (!this._eof() && this._peek() === '>') {
            this._consume('>')
            this._push(TokenType.TAG_CLOSE, '>', this._startPos())
        }
    }

    _readComment() {
        const start = this._startPos()
        this._consume('<!--')
        let content = ''
        while (!this._eof() && !this._lookingAt('-->')) {
            content += this._advance()
        }
        if (this._eof()) {
            throwError(ErrorCode.MALFORMED_COMMENT, createLoc(start, this._startPos()), this.source)
        }
        this._consume('-->')
        this._push(TokenType.COMMENT, content.trim(), start)
    }

    _readMustache() {
        const openStart = this._startPos()
        this._push(TokenType.MUSTACHE_OPEN, '{{', openStart)

        const exprStart = this._startPos()
        let expr = ''
        let depth = 0
        let inStr = false
        let strChar = null

        while (!this._eof()) {
            const ch = this._peek()

            if (inStr) {
                if (ch === '\\') {
                    expr += this._advance()
                    expr += this._advance()
                    continue
                }
                if (ch === strChar) {
                    inStr = false
                }
                expr += this._advance()
                continue
            }

            if (ch === '"' || ch === "'" || ch === '`') {
                inStr = true
                strChar = ch
                expr += this._advance()
                continue
            }
            if (ch === '{') {
                depth++;
                expr += this._advance()
                continue
            }
            if (ch === '}') {
                if (depth > 0) {
                    depth--
                    expr += this._advance()
                    continue
                }
                if (this._peekAt(1) === '}') {
                    this._push(TokenType.MUSTACHE_EXPR, expr.trim(), exprStart)
                    const closeStart = this._startPos()
                    this._consume('}}')
                    this._push(TokenType.MUSTACHE_CLOSE, '}}', closeStart)
                    return
                }
            }

            expr += this._advance()
        }

        throwError(ErrorCode.UNTERMINATED_MUSTACHE, createLoc(openStart, this._startPos()), this.source, expr)
    }

    _readInTag() {
        this._skipWhitespace()
        if (this._eof()) return

        if (this._match('/>')) {
            this._push(TokenType.TAG_SELF_CLOSE, '/>', this._startPos())
            this._mode = 'text'
            return
        }

        if (this._peek() === '>') {
            this._consume('>')
            this._push(TokenType.TAG_CLOSE, '>', this._startPos())
            this._mode = 'text'
            return
        }

        const start = this._startPos()
        const name = this._readAttrName()
        if (!name) return

        this._push(TokenType.ATTR_NAME, name, start)
        this._skipWhitespace()

        if (this._peek() === '=') {
            this._consume('=')
            this._push(TokenType.ATTR_EQUALS, '=', this._startPos())
            this._skipWhitespace()
            this._readAttrValue()
        }
    }

    _readAttrName() {
        let name = ''

        if (this._peek() === '@') {
            name += this._advance()
        }
        else if (this._peek() === ':') {
            name += this._advance()
        }

        while (!this._eof()) {
            const ch = this._peek()
            if (/[a-zA-Z0-9\-_.:]/.test(ch)) {
                name += this._advance()
            } 
            else 
                break
        }

        return name || null
    }

    _readAttrValue() {
        if (this._eof()) return
        const quote = this._peek()
        if (quote !== '"' && quote !== "'") {
            const start = this._startPos()
            let val = ''
            while (!this._eof() && !/[\s>]/.test(this._peek())) {
                val += this._advance()
            }
            this._push(TokenType.ATTR_VALUE, val, start)
            return
        }

        this._consume(quote)
        const start = this._startPos()
        let val = ''

        while (!this._eof()) {
            const ch = this._peek()
            if (ch === quote) {
                this._consume(quote)
                break
            }
            if (ch === '\\') {
                this._advance()
                val += this._advance()
                continue
            }
            val += this._advance()
        }

        this._push(TokenType.ATTR_VALUE, val, start)
    }

    _readTagName() {
        let name = ''
        while (!this._eof() && /[a-zA-Z0-9\-_.]/.test(this._peek())) {
            name += this._advance()
        }
        return name
    }

    _isTagNameStart(ch) {
        return ch != null && /[a-zA-Z_]/.test(ch)
    }

    _eof() {
        return this.pos >= this.source.length
    }

    _peek() {
        return this.source[this.pos]
    }

    _peekAt(n) {
        return this.source[this.pos + n]
    }

    _lookingAt(str) {
        return this.source.startsWith(str, this.pos)
    }

    _match(str) {
        if (this._lookingAt(str)) {
            for (const ch of str) 
                this._advance()
            return true
        }
        return false
    }

    _consume(expected) {
        for (const ch of expected) {
            if (this.source[this.pos] !== ch) {
                throwError(
                    ErrorCode.UNEXPECTED_CHAR,
                    createLoc(this._startPos(), this._startPos()),
                    this.source,
                    this.source[this.pos]
                )
            }
            this._advance()
        }
    }

    _advance() {
        const ch = this.source[this.pos++]
        if (ch === '\n') {
            this.line++;
            this.col = 1
        } 
        else {
            this.col++
        }
        return ch
    }

    _skipWhitespace() {
        while (!this._eof() && /\s/.test(this._peek())) this._advance()
    }

    _startPos() {
        return createPos(this.line, this.col, this.pos)
    }

    _push(type, value, start) {
        this.tokens.push({
            type,
            value,
            loc: createLoc(start, this._startPos(), value)
        })
    }

    _pushAt(type, value, start) {
        this._push(type, value, start)
    }
}

File: framework/core/rust-compiler/src/lexer.rs
use crate::args;
use crate::ast::{Loc, Pos};
use crate::errors::{CompileError, ErrorCode};
use serde::Serialize;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
pub enum TokenType {
    TagOpen,
    TagOpenClose,
    TagClose,
    TagSelfClose,
    Comment,
    AttrName,
    AttrEquals,
    AttrValue,
    Text,
    MustacheOpen,
    MustacheExpr,
    MustacheClose,
    Eof,
}

#[derive(Debug, Clone, Serialize)]
pub struct Token {
    pub ttype: TokenType,
    pub value: String,
    pub loc: Loc,
}

#[derive(PartialEq, Eq, Clone, Copy)]
enum Mode {
    Text,
    Tag,
}

pub fn tokenize(source: &str) -> Result<Vec<Token>, CompileError> {
    let mut lexer = Lexer::new(source);
    lexer.tokenize()
}

struct Lexer {
    chars: Vec<char>,
    source: String,
    tokens: Vec<Token>,
    pos: usize,
    line: usize,
    col: usize,
    mode: Mode,
}

impl Lexer {
    fn new(source: &str) -> Self {
        Lexer {
            chars: source.chars().collect(),
            source: source.to_string(),
            tokens: Vec::new(),
            pos: 0,
            line: 1,
            col: 1,
            mode: Mode::Text,
        }
    }

    fn tokenize(&mut self) -> Result<Vec<Token>, CompileError> {
        while !self.eof() {
            match self.mode {
                Mode::Text => self.read_text()?,
                Mode::Tag => self.read_in_tag()?,
            }
        }
        let start = self.start_pos();
        self.push(TokenType::Eof, String::new(), start);
        Ok(std::mem::take(&mut self.tokens))
    }

    fn read_text(&mut self) -> Result<(), CompileError> {
        if self.looking_at("<!--") {
            return self.read_comment();
        }
        if self.looking_at("</") {
            return self.read_closing_tag();
        }
        if self.peek() == Some('<') && self.is_tag_name_start(self.peek_at(1)) {
            return self.read_open_tag();
        }
        if self.looking_at("{{") {
            return self.read_mustache();
        }

        let start = self.start_pos();
        let mut text = String::new();
        while !self.eof() {
            if self.peek() == Some('<') || self.looking_at("{{") {
                break;
            }
            text.push(self.advance());
        }
        if !text.is_empty() {
            self.push(TokenType::Text, text, start);
        }
        Ok(())
    }

    fn read_open_tag(&mut self) -> Result<(), CompileError> {
        let start = self.start_pos();
        self.consume("<")?;
        let name = self.read_tag_name();
        self.push(TokenType::TagOpen, name, start);
        self.mode = Mode::Tag;
        Ok(())
    }

    fn read_closing_tag(&mut self) -> Result<(), CompileError> {
        let start = self.start_pos();
        self.consume("</")?;
        let name = self.read_tag_name();
        self.push(TokenType::TagOpenClose, name, start);
        self.skip_whitespace();
        if !self.eof() && self.peek() == Some('>') {
            let close_start = self.start_pos();
            self.consume(">")?;
            self.push(TokenType::TagClose, ">".into(), close_start);
        }
        Ok(())
    }

    fn read_comment(&mut self) -> Result<(), CompileError> {
        let start = self.start_pos();
        self.consume("<!--")?;
        let mut content = String::new();
        while !self.eof() && !self.looking_at("-->") {
            content.push(self.advance());
        }
        if self.eof() {
            return Err(CompileError::new(
                ErrorCode::MalformedComment,
                Loc::new(start, self.start_pos()),
                &self.source,
                args!(),
            ));
        }
        self.consume("-->")?;
        self.push(TokenType::Comment, content.trim().to_string(), start);
        Ok(())
    }

    fn read_mustache(&mut self) -> Result<(), CompileError> {
        let open_start = self.start_pos();
        self.consume("{{")?;
        self.push(TokenType::MustacheOpen, "{{".into(), open_start);

        let expr_start = self.start_pos();
        let mut expr = String::new();
        let mut depth = 0i32;
        let mut in_str = false;
        let mut str_char: Option<char> = None;

        while !self.eof() {
            let ch = self.peek().unwrap();

            if in_str {
                if ch == '\\' {
                    expr.push(self.advance());
                    if !self.eof() {
                        expr.push(self.advance());
                    }
                    continue;
                }
                if Some(ch) == str_char {
                    in_str = false;
                }
                expr.push(self.advance());
                continue;
            }

            if ch == '"' || ch == '\'' || ch == '`' {
                in_str = true;
                str_char = Some(ch);
                expr.push(self.advance());
                continue;
            }
            if ch == '{' {
                depth += 1;
                expr.push(self.advance());
                continue;
            }
            if ch == '}' {
                if depth > 0 {
                    depth -= 1;
                    expr.push(self.advance());
                    continue;
                }
                if self.peek_at(1) == Some('}') {
                    self.push(TokenType::MustacheExpr, expr.trim().to_string(), expr_start);
                    let close_start = self.start_pos();
                    self.consume("}}")?;
                    self.push(TokenType::MustacheClose, "}}".into(), close_start);
                    return Ok(());
                }
            }

            expr.push(self.advance());
        }

        Err(CompileError::new(
            ErrorCode::UnterminatedMustache,
            Loc::new(open_start, self.start_pos()),
            &self.source,
            args!(expr),
        ))
    }

    fn read_in_tag(&mut self) -> Result<(), CompileError> {
        self.skip_whitespace();
        if self.eof() {
            return Ok(());
        }

        if self.matches("/>") {
            self.push(TokenType::TagSelfClose, "/>".into(), self.start_pos());
            self.mode = Mode::Text;
            return Ok(());
        }

        if self.peek() == Some('>') {
            self.consume(">")?;
            self.push(TokenType::TagClose, ">".into(), self.start_pos());
            self.mode = Mode::Text;
            return Ok(());
        }

        let start = self.start_pos();
        let name = self.read_attr_name();
        let name = match name {
            Some(n) => n,
            None => return Ok(()),
        };

        self.push(TokenType::AttrName, name, start);
        self.skip_whitespace();

        if self.peek() == Some('=') {
            self.consume("=")?;
            self.push(TokenType::AttrEquals, "=".into(), self.start_pos());
            self.skip_whitespace();
            self.read_attr_value();
        }

        Ok(())
    }

    fn read_attr_name(&mut self) -> Option<String> {
        let mut name = String::new();

        if self.peek() == Some('@') || self.peek() == Some(':') {
            name.push(self.advance());
        }

        while !self.eof() {
            let ch = self.peek().unwrap();
            if ch.is_ascii_alphanumeric() || "-_.:".contains(ch) {
                name.push(self.advance());
            } else {
                break;
            }
        }

        if name.is_empty() {
            None
        } else {
            Some(name)
        }
    }

    fn read_attr_value(&mut self) {
        if self.eof() {
            return;
        }
        let quote = self.peek().unwrap();
        if quote != '"' && quote != '\'' {
            let start = self.start_pos();
            let mut val = String::new();
            while !self.eof() {
                let ch = self.peek().unwrap();
                if ch.is_whitespace() || ch == '>' {
                    break;
                }
                val.push(self.advance());
            }
            self.push(TokenType::AttrValue, val, start);
            return;
        }

        self.advance();
        let start = self.start_pos();
        let mut val = String::new();

        while !self.eof() {
            let ch = self.peek().unwrap();
            if ch == quote {
                self.advance();
                break;
            }
            if ch == '\\' {
                self.advance();
                if !self.eof() {
                    val.push(self.advance());
                }
                continue;
            }
            val.push(self.advance());
        }

        self.push(TokenType::AttrValue, val, start);
    }

    fn read_tag_name(&mut self) -> String {
        let mut name = String::new();
        while !self.eof() {
            let ch = self.peek().unwrap();
            if ch.is_ascii_alphanumeric() || "-_.".contains(ch) {
                name.push(self.advance());
            } else {
                break;
            }
        }
        name
    }

    fn is_tag_name_start(&self, ch: Option<char>) -> bool {
        matches!(ch, Some(c) if c.is_ascii_alphabetic() || c == '_')
    }


    fn eof(&self) -> bool {
        self.pos >= self.chars.len()
    }

    fn peek(&self) -> Option<char> {

        self.chars.get(self.pos).copied()
    }

    fn peek_at(&self, n: usize) -> Option<char> {

        self.chars.get(self.pos + n).copied()
    }

    fn looking_at(&self, s: &str) -> bool {
        let s_chars: Vec<char> = s.chars().collect();
        if self.pos + s_chars.len() > self.chars.len() {
            return false;
        }
        self.chars[self.pos..self.pos + s_chars.len()] == s_chars[..]
    }

    fn matches(&mut self, s: &str) -> bool {
        if self.looking_at(s) {
            for _ in s.chars() {
                self.advance();
            }
            true
        } else {
            false
        }
    }

    fn consume(&mut self, expected: &str) -> Result<(), CompileError> {
        for ch in expected.chars() {
            if self.chars.get(self.pos) != Some(&ch) {
                let found = self
                    .chars
                    .get(self.pos)
                    .map(|c| c.to_string())
                    .unwrap_or_default();
                return Err(CompileError::new(
                    ErrorCode::UnexpectedChar,
                    Loc::new(self.start_pos(), self.start_pos()),
                    &self.source,
                    args!(found),
                ));
            }
            self.advance();
        }
        Ok(())
    }

    fn advance(&mut self) -> char {
        let ch = self.chars[self.pos];
        self.pos += 1;
        if ch == '\n' {
            self.line += 1;
            self.col = 1;
        } else {
            self.col += 1;
        }
        ch
    }

    fn skip_whitespace(&mut self) {
        while !self.eof() && self.peek().map(|c| c.is_whitespace()).unwrap_or(false) {
            self.advance();
        }
    }

    fn start_pos(&self) -> Pos {
        Pos::new(self.line, self.col, self.pos)
    }

    fn push(&mut self, ttype: TokenType, value: String, start: Pos) {
        let loc = Loc::new(start, self.start_pos());
        self.tokens.push(Token { ttype, value, loc });
    }
}